ipv8/attestation/communication_manager.py

```diff
@@ -0,0 +1,258 @@
+                               (self.public_key_bin, self.public_key_bin, self.public_key_bin))
+        database.commit()
+
+        return attestation_hashes
+
+    def _drop_attestation_table_data(self, attestation_hashes):
+        """
+        Remove all attestation data (claim based keys and ZKP blobs) by list of attestation hashes.
+
+        :param attestation_hashes: hashes to remove
+        :type attestation_hashes: [database_blob]
+        :returns: None
+        """
+        if not attestation_hashes:
+            return
+
+        self.attestation_overlay.database.execute(("DELETE FROM %s" % self.attestation_overlay.database.db_name)
+                                                  + " WHERE hash IN ("
+                                                  + ", ".join(c for c in "?" * len(attestation_hashes))
+                                                  + ")",
+                                                  attestation_hashes)
+        self.attestation_overlay.database.commit()
+
+    def remove(self):
+        """
+        Remove this pseudonym from existence.
+        """
+        self._drop_attestation_table_data(self._drop_identity_table_data())
+        self.attestation_requests.clear()
+
+    def get_my_attributes(self):
+        return self.get_attributes(self.identity_overlay.my_peer)
+
+    def get_attributes(self, peer):
+        pseudonym = self.identity_overlay.identity_manager.get_pseudonym(peer.public_key)
+        out = {}
+        for credential in pseudonym.get_credentials():
+            attestations = list(credential.attestations)
+            attesters = []
+            for attestation in attestations:
+                attesters.append(self.identity_overlay.identity_manager.database.get_authority(attestation))
+            attribute_hash = pseudonym.tree.elements[credential.metadata.token_pointer].content_hash
+            json_metadata = json.loads(credential.metadata.serialized_json_dict)
+            out[attribute_hash] = (json_metadata["name"], json_metadata, attesters)
+        return out
+
+    def request_attestation(self, peer, attribute_name, id_format, metadata):
+        key = self.attestation_overlay.get_id_algorithm(id_format).generate_secret_key()
+        metadata.update({"id_format": id_format})
+        self.attestation_metadata[(self.identity_overlay.my_peer, attribute_name)] = metadata
+        self.attestation_overlay.request_attestation(peer, attribute_name, key, metadata)
+
+    def attest(self, peer, attribute_name, value):
+        outstanding = self.attestation_requests.pop((peer, attribute_name))
+        outstanding[0].set_result(value)
+
+    def import_blob(self, attribute_name, id_format, metadata, value):
+        metadata.update({"id_format": id_format})
+        self.attestation_overlay.dump_blob(attribute_name, id_format, value, metadata)
+
+    def allow_verification(self, peer, attribute_name):
+        outstanding = self.verify_requests.pop((peer, attribute_name))
```

Comment:
What if verification is not allowed, is there some kind of timeout for this? Otherwise, the message handler that awaits this `Future` could end up waiting indefinitely.

Reply:
A user should be able to answer at any point in time he likes, so there is no timeout. However, having a "disallow" may indeed be a good idea.